hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java

```diff
@@ -8183,16 +8183,22 @@ else if ( getLoadQueryInfluencers().hasEnabledFetchProfiles() ) {
 						final TableGroup actualTableGroup = joinedTableGroup instanceof PluralTableGroup ?
 								( (PluralTableGroup) joinedTableGroup ).getElementTableGroup() :
 								joinedTableGroup;
-						final MappingType entityMappingType = actualTableGroup == null
+						final MappingType mappingType = actualTableGroup == null
 								? null
 								: actualTableGroup.getModelPart().getPartMappingType();
-						if ( entityMappingType instanceof EntityMappingType ) {
+						if ( mappingType instanceof EntityMappingType ) {
+							final EntityMappingType entityMappingType = (EntityMappingType) mappingType;
 							registerEntityNameUsage(
 									actualTableGroup,
 									EntityNameUse.PROJECTION,
-									( (EntityMappingType) entityMappingType ).getEntityName(),
+									entityMappingType.getEntityName(),
 									true
 							);
+							if ( entityMappingType.getSuperMappingType() != null ) {
+								// A joined table group was created by an enabled entity graph or fetch profile,
+								// and it's of an inheritance subtype, so we should apply the discriminator
+								entityMappingType.applyDiscriminator( null, null, actualTableGroup, this );
```

Comment:
Shouldn't we do something similar for `org.hibernate.loader.ast.internal.LoaderSelectBuilder#createFetchableConsumer` around line 1000?